### PR TITLE
packaging: use setx /m to register wazero instead

### PR DESCRIPTION
**Pull request by evacchi**
Addresses #1374.

The most important issue with this strategy is that it has
the potential to break the end-user PATH variable if for some
reason the combined length is longer than 1024 chars.

The reason the current strategy is not working are at least two:

1. `%PATH%` is not expanded in:

    <CustomAction Id="ChangePath"
        ExeCommand="PATH &quot;%PATH%;[INSTALLDIR]&quot;"
        Property="setx" ...

    The "correct" syntax seems to be `[%PATH]` instead of `%PATH%`

2. SETX is not working(*) inside MSI (at least on my Win11 machine)
   unless we use the flag `/M` which modifies the env variable
   at the system level.

Furthermore:

- `SETX PATH "C:\Program Files\wazero;%PATH%"` expands `%PATH%`
  AND *truncates* it to 1024 chars. Because of this, *appending* may
  be safer: worst case, it truncates the path you have appended

- However, for some reason I couldn't make it work from within
  the MSI installer, so I still had to prepend...

(*) I assume it's running as a different user than %currentuser% 
and SYSTEM; maybe "Local Service"?

Signed-off-by: Edoardo Vacchi <evacchi@users.noreply.github.com>

### CODE REVIEW

**Review by mathetake — COMMENTED** [review]
can this be testable in [verify_msi.cmd](https://github.com/tetratelabs/wazero/blob/main/packaging/msi/verify_msi.cmd) which is executed in the [pre_release_test](https://github.com/tetratelabs/wazero/blob/main/.github/workflows/release.yaml#L145) CI?

**Comment by evacchi** [comment]
Will do. I am also further investigating a different approach (using the actual WIX tool chain instead)

**Comment by evacchi** [comment]
I have updated the wix file to v4 (it may not be strictly required TBF) and actually used the `Environment` section which seems to be working as expected (it shouldn't be a surprise, since I think this is the official tool). If we do go this route it'll require proper setup of the CI (but I think there are a bunch of GHA that may be used) see WIP https://github.com/evacchi/wazero/commit/464a9e37ca6905e3ab960481870d2350a5f45581#diff-217b40bac6395c4cd76179d71ebc5cda14719c5fec343444884d6faff2e61b2fR98

**Review by codefromthecrypt — APPROVED** [review]
I'm fine either way (this or switching to v4, or in steps both)

for verification, somehow it looked like we were verifying before, so maybe that's a different task to fix the false positive, but ideally done alongside the fix.

**Comment by evacchi** [comment]
yes, I was just checking it and apparently we already verifying this. The issue may be that GHA is running as a privileged user so the problem doesn't present itself.

wix v3 is already present in the windows-2022 runner, it could be configured under Ubuntu but it requires to install wine and some fiddling... 🤔 so either we break the build into 2 steps or we figure out how to configure wix v3 under the Ubuntu runner... 🤔

**Comment by codefromthecrypt** [comment]
just a thought.. if CI was working because the result path <1024 chars, maybe the CI step could be configured with a really long PATH (even if invalid entries) to force the issue..

**Comment by evacchi** [comment]
in fact, in the meantime I have tested this smaller patch under the native wix3 toolchain and it seems to work https://github.com/evacchi/wazero/commit/wix3-native-build

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -141,6 +141,10 @@ jobs:
       - name: Test Windows Installer
         if: runner.os == 'Windows'
         run: |
+          SET GARBAGE="12345678901234567890123456789012345678901234567890123456789012345678901234567890"
+          SET GARBAGE="%GARBAGE%%GARBAGE%%GARBAGE%%GARBAGE%%GARBAGE%%GARBAGE%%GARBAGE%%GARBAGE%"
+          SET GARBAGE="%GARBAGE%%GARBAGE%"
+          SETX PATH "%PATH%;%GARBAGE"
```
Comment (evacchi) [inline .github/workflows/release.yaml:147]: the cmd script is not failing with non-zero (I'll try and see how to make it actually fail), but this is indeed breaking PATH:

```
WARNING: The data being saved is truncated to 1024 characters.

SUCCESS: Specified value was saved.
installing "dist\wazero_8df8b975d660774e980afe9381df559e8e871167_windows_amd64.msi"
Refreshing environment variables from registry for cmd.exe. Please wait...Finished..
ensuring wazero was installed
8df8b975d660774e980afe9381df559e8e871[16](https://github.com/tetratelabs/wazero/actions/runs/4755047545/jobs/8448708184#step:6:17)7
uninstalling wazero
ensuring wazero was uninstalled
'wazero' is not recognized as an internal or external command,
operable program or batch file.
```

Comment (codefromthecrypt) [reply .github/workflows/release.yaml:147]: thanks for your tenacity again. you seem to be a magnet for subtlety broken things :)

**Comment by evacchi** [comment]
I have created https://github.com/tetratelabs/wazero/pull/1385 because I wanted to show both approaches. If I may explain visually:



In this PR:
- I could not manage to make the CI fail reliably when the %PATH% is corrupted. For that matter, I am not sure I was able to corrupt %PATH%
- The only approach that WorksForMe™️ is to _prepend_ the install path to the *system* path using `SETX /M`, which for long PATH env may push one of the valid elements of the list off the 1024 boundary, effectively corrupting the system PATH. Notice that I am not an expert, so I am just assuming.
- Another approach may be editing the registry directly; I didn't try this.

in PR #1385:
- msitools do not support the Environment directive, so we can try the "official" tools
- I was not able to run native WiX under macOS/Linux. They are not supported because they use native platform libraries. In theory there are ways to run it under wine: it didn't work for our project, in my tests (tried that too!)
- I am running the entire `make dist` step under on the `windows-2022` runner so that we can use native WiX Tools v3 
- alternative: run only `make dist "msi"` under a Windows runner

**Comment by codefromthecrypt** [comment]
Thanks for the writeup. I think it is ok for us to run the release on windows until someone can compile the official tools to wasm, so we can run them on linux.

I was hoping msitools would progress, but the environment thing there went no-where.

As long as @mathetake is ok running on windows, I think we should go with PR https://github.com/tetratelabs/wazero/pull/1385. Maybe you can build an MSI on that and make sure the original person who had the problem is able to use it (e.g. attach the MSI to the PR or the issue leading to it, zipping if necessary)

**Comment by evacchi** [comment]
closing in favor of https://github.com/tetratelabs/wazero/pull/1385

**Pull request closed without merging**